### PR TITLE
Bug 1148442 - Add support for the W3C icons property to <gaia-grid>

### PR DESCRIPTION
**Pull request by benfrancis**

### CODE REVIEW

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5045e93201574d67a9b5dd3c03bc492201dcb43a

File: shared/elements/gaia_grid/js/items/grid_item.js [hunk 1]
```diff
@@ -197,6 +197,18 @@
       if (!choices) {
         return this.defaultIcon;
       }
+      var icon;
+      var maxSize = this.grid.layout.gridMaxIconSize; // The goal size
+
+      // Check for W3C web manifest format for icons
+      if (choices.constructor === Array) {
```
Comment (KevinGrandon) [inline shared/elements/gaia_grid/js/items/grid_item.js]: Use the ```Array.isArray()``` check here instead.

File: shared/js/web_manifest_helper.js [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+  /**
+   * Get Manifest.
+   *
+   * @param {String} url Manifest URL
+   * @returns {Promise} A promise of a response containing manifest data.
+   */
+  var getManifest = function(url) {
+    return new Promise(function(resolve, reject) {
+      var xhr = new XMLHttpRequest({mozSystem: true});
```
Comment (KevinGrandon) [inline shared/js/web_manifest_helper.js]: I want to use the ```LazyLoader.getJSON``` method here instead. The only thing I don't think it supports yet is ```setRequestHeader```. We might need to extend it to do so if this is required.

**Comment by KevinGrandon** [comment]
Do we need to remove the web_manifest_helper.js file from the bookmarks app?

File: shared/elements/gaia_grid/js/items/grid_item.js (hunk 1 above)
Comment (benfrancis) [inline shared/elements/gaia_grid/js/items/grid_item.js]: Done. (I only did it this way because I read it was faster).

**Comment by benfrancis** [comment]
> Do we need to remove the web_manifest_helper.js file from the bookmarks app?

Ah, sorry, I somehow managed to miss off the deleted files from the commit, I suspect that's what was causing the test failures too.

web_manifest_helper.js has actually already been code reviewed.

If WebManifestHelper.getManifest() uses LazyLoader.getJSON() then that makes LazyLoader a dependency of GaiaGrid which means it needs loading in about 7 more HTML files. I know this code wraps an XHR in a Promise similar to getJSON but my understanding was that LazyLoader is really meant for lazy loading parts of an app, whereas this is for fetching a specific type of remote resource. Depending on the outcome of bug 1091786 we may also need to add Content Type to the response to filter out manifests which don't specify the correct Content Type. This would be a weird addition to getJSON as local resources probably never have a Content Type.

Let me know if you still think this should use LazyLoader, I personally think this usage is a bit different.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=933179e03bc82b17581c057c74626e8de668847e

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8f844a52b9b434f8d8801f94a87624678878ac1e

**Pull request merged**